graph/common/Encoding.java

```diff
@@ -400,13 +401,27 @@ public Infix infix() {
                 pair(DATETIME, set(DATETIME))
         );
 
-        private static final Map<ValueType, Set<ValueType>> COMPARABLES = map(
-                pair(OBJECT, set(OBJECT)),
-                pair(BOOLEAN, set(BOOLEAN)),
-                pair(LONG, set(LONG, DOUBLE)),
-                pair(DOUBLE, set(LONG, DOUBLE)),
-                pair(STRING, set(STRING)),
-                pair(DATETIME, set(DATETIME))
+        private static final Map<ValueType, Compare> COMPARABLES = map(
+                pair(OBJECT, new Compare(map(
+                        pair(OBJECT, Comparator.comparing(val -> val, (a, b) -> 0))
+                ))),
+                pair(BOOLEAN, new Compare(map(
+                        pair(BOOLEAN, Comparator.comparing(bool -> (boolean) bool, Boolean::compareTo))
+                ))),
+                pair(LONG, new Compare(map(
+                        pair(LONG, Comparator.comparing(l -> (long) l, Long::compareTo)),
+                        pair(DOUBLE, Comparator.comparing(val -> (double) val, Double::compareTo))
+                ))),
+                pair(DOUBLE, new Compare(map(
+                        pair(LONG, Comparator.comparing(d -> (double) d, Double::compareTo)),
+                        pair(DOUBLE, Comparator.comparing(d -> (double) d, Double::compareTo))
+                ))),
+                pair(DATETIME, new Compare(map(
+                        pair(DATETIME, Comparator.comparing(d -> (LocalDateTime) d, LocalDateTime::compareTo))
+                ))),
+                pair(STRING, new Compare(map(
+                        pair(STRING, Comparator.comparing(s -> (String) s, String::compareToIgnoreCase))
+                )))
```

Comment:
we now define the comparables AND the comparator to use when comparing to a particular type in one place, rather than scattering the implementation around